### DISCUSSION TITLE
Inline-able plain enum

### DISCUSSION THREAD

**Discussion by Jack-Works** [Jack-Works/proposal-enum#4] [General]
[Flow.js enum](https://flow.org/en/docs/enums/defining-enums/) is designed inline-friendly. [TypeScript has const enum](https://www.typescriptlang.org/docs/handbook/enums.html#const-enums) that can be inlined by the TypeScript compiler.

Inline enum is important in the performance/size-sensitive case.

## Example in TypeScript

```ts
const enum SyntaxKind {
    Statement, Expression
}
const node = [ SyntaxKind.Expression, SyntaxKind.Statement ]
```

The code above will be compiled into:

```js
const node = [1 /* Expression */, 0 /* Statement */];
```

## Problem

Should we constraint the design of this proposal to keep it inline-able when necessary?

To keep it inline-able, the following features are not being included (or maybe we can have "partial" inline-able?)

- Full version of #11
- Full version of #12
- #5
- #9 does not inline well
- #8 without special syntax for symbols because constructing a new symbol need access to the Symbol constructor

**Comment by ljharb** [comment]
I don’t understand this requirement. JS doesn’t have a compiler; the spec - and thus the proposal - can’t constrain what build tools can or can not do.

**Reply by Jack-Works** [reply]
This is intended to make it more statically analyzable by engines and toolchains.

**Reply by ljharb** [reply]
This requirement would eliminate Symbols, which seems like it’s a nonstarter to me.

**Reply by Jack-Works** [reply]
```
enum A of symbols { A, B, C }
```

It can support symbols if we do not need Heterogeneous Enum (#8)

**Reply by ljharb** [reply]
If i can’t make an enum of well-known symbols, or existing ones, i don’t think it’s met the usefulness bar.

**Reply by legendecas** [reply]
@Jack-Works I'm confused, why supporting symbols conflicts with the heterogeneous values?

**Reply by Jack-Works** [reply]
> If i can’t make an enum of well-known symbols, or existing ones, i don’t think it’s met the usefulness bar.

That's a reasonable argument to me. So we should give up this design constraint?

**Comment by rbuckton** [comment]
Inlining is tricky. TS doesn't inline a regular `enum` as its possible the package you are depending on for that enum could renumber its values, which might not be caught if you're not using a pinned version of the dependency. `const` enums are faster at runtime, but aren't tolerant to version changes.

I don't think its necessary to define `enum` to be inlinable in the specification. That's something a host like v8 could possibly do internally if there was enough benefit to be gained by the optimization.

**Reply by Jack-Works** [reply]
> I don't think its necessary to define enum to be inlinable in the specification. That's something a host like v8 could possibly do internally if there was enough benefit to be gained by the optimization.

I just wonder if we should design it in a more inline-friendly way (like flow.js does). It doesn't mean the host/compiler must inline them.
